server/src/main/java/org/apache/kafka/server/share/ShareSessionCache.java

```diff
@@ -0,0 +1,174 @@
+ * <p>
+ * The ShareSessionCache is thread-safe because all of its methods are synchronized.
+ * Note that individual share sessions have their own locks which are separate from the
+ * ShareSessionCache lock.  In order to avoid deadlock, the ShareSessionCache lock
+ * must never be acquired while an individual ShareSession lock is already held.
+ */
+public class ShareSessionCache {
+    private final int maxEntries;
+    private final long evictionMs;
+    private long numPartitions = 0;
+
+    // A map of session key to ShareSession.
+    private Map<ShareSessionKey, ShareSession> sessions = new HashMap<>();
+
+    // Maps last used times to sessions.
+
+    private TreeMap<LastUsedKey, ShareSession> lastUsed = new TreeMap<>();
+
+    // Visible for testing
+    public TreeMap<LastUsedKey, ShareSession> lastUsed() {
+        return lastUsed;
+    }
+
+    public ShareSessionCache(int maxEntries, long evictionMs) {
+        this.maxEntries = maxEntries;
+        this.evictionMs = evictionMs;
+    }
+
+    /**
+     * Get a session by session key.
+     *
+     * @param key The share session key.
+     * @return The session, or None if no such session was found.
+     */
+    public ShareSession get(ShareSessionKey key) {
+        synchronized (this) {
+            return sessions.getOrDefault(key, null);
+        }
+    }
+
+    /**
+     * Get the number of entries currently in the share session cache.
+     */
+    public int size() {
+        synchronized (this) {
+            return sessions.size();
+        }
+    }
+
+    public long totalPartitions() {
+        synchronized (this) {
+            return numPartitions;
+        }
+    }
+
+    public ShareSession remove(ShareSessionKey key) {
+        synchronized (this) {
+            ShareSession session = get(key);
+            if (session != null)
+                return remove(session);
+            return null;
+        }
+    }
+
+    /**
+     * Remove an entry from the session cache.
+     *
+     * @param session The session.
+     * @return The removed session, or None if there was no such session.
+     */
+    public ShareSession remove(ShareSession session) {
+        synchronized (this) {
+            synchronized (session) {
+                lastUsed.remove(session.lastUsedKey());
+            }
+            ShareSession removeResult = sessions.remove(session.key());
+            if (removeResult != null) {
+                numPartitions = numPartitions - session.cachedSize();
+            }
+            return removeResult;
+        }
+    }
+
+    /**
+     * Update a session's position in the lastUsed tree.
+     *
+     * @param session  The session.
+     * @param now      The current time in milliseconds.
+     */
+    public void touch(ShareSession session, long now) {
+        synchronized (this) {
+            synchronized (session) {
+                // Update the lastUsed map.
+                lastUsed.remove(session.lastUsedKey());
```

Comment:
Do we need to remove here? it will replaced at line no:126

Reply:
Yes, we need to remove here because the position of the element in the `TreeMap` has to be updated as well which is important when we have an element to evict using the `tryEvict` policy. Otherwise, "not the most stale session" can be removed when trying to evict from the cache.